### PR TITLE
Update all the E2E should_fail tests to verify their output.

### LINKED ISSUE #723: Refactor the E2E test driver.

### LINKED ISSUE #1580: Update E2E test harness to verify specific failures.
I think we've talked about this but I can't find a specific issue tracking it -- the `should_fail` tests should only 'pass' when the failures (errors and/or warnings) are specifically what we expect.  At the moment the projects could fail to build for any reason and the tests would still pass.

### PR DESCRIPTION
**Pull request by otrho**
Using the `FileCheck` crate it can now pattern match against the compiler output to be sure the errors and/or warnings are exactly what we expect.

Closes #723.
Closes #1580.

### CODE REVIEW

**Comment by kayagokalp** [comment]
I am in the middle of merging some stuff and will look into this as soon as it finishes 👍

**Comment by mitchmindtree** [comment]
Okydoke @otrho, should be solved via #2090.

File: test/src/e2e_vm_tests/test_programs/should_fail/excess_fn_arguments/test.toml [hunk 1]
```diff
@@ -1 +1,5 @@
-category = "fail"
+# This test should be a "fail" but it's currently passing.
```
Comment (mitchmindtree) [inline test/src/e2e_vm_tests/test_programs/should_fail/excess_fn_arguments/test.toml:1]: I wonder we should have a flag or something that indicates that the test is known not to pass right now but _should_ in the future? Then we could potentially print a list of known not-yet-passing tests at the end of the test output as a reminder? That said I guess having an issue for these is probably enough. Either way, not a concern of this PR!

Comment (mitchmindtree) [reply test/src/e2e_vm_tests/test_programs/should_fail/excess_fn_arguments/test.toml:1]: Ah I just noticed we have a `disabled` category - guess we could use that?

**Review by mitchmindtree — APPROVED** [review]
Wicked, TIL about `filecheck`.

Thanks loads for tackling these test issues @otrho, looking so much better!

File: test/src/e2e_vm_tests/test_programs/should_fail/excess_fn_arguments/test.toml (hunk 1 above)
Comment (otrho) [reply test/src/e2e_vm_tests/test_programs/should_fail/excess_fn_arguments/test.toml:1]: Yar, it prints the number of disabled tests at the end, I guess we could keep an eye on that.

Having an actual GitHub issue for things that should change in the future is a bit more robust.  I reckon having a flag in the suite would just get forgotten too easily.

**Review by tritao — APPROVED** [review]
Lgtm

**Pull request merged**